### ISSUE TITLE
Design cyclic reactive dataflow IR for inference metadata

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#2337]
## Summary

Explore replacing the authored `sequence` / `branch` / `loop` / `invoke` / `emit` workflow syntax with a uniform cyclic, typed dataflow graph. Components and state are nodes, typed values are edges, and feedback through a state node expresses recurrence.

The authored graph should remain portable and semantic. At load time, it is validated and compiled into structured regions, state transactions, and a static execution plan; the runtime must not interpret the generic graph in the per-token hot path.

## Motivation

The current workflow IR explicitly encodes structured control flow. A graph-first representation may provide a smaller and more composable authoring model:

- a component node replaces `invoke`;
- ordinary data/effect dependencies replace `sequence`;
- predicate-driven switch/gate components replace `branch`;
- workflow output bindings replace `emit`;
- a cycle through a state node replaces `loop`.

This must not remove the semantics required to make cycles deterministic: initialization, delay, firing/reaction boundaries, termination, state ownership, effects, and commit/rollback behavior still need precise definitions.

### Nodes and edges

- Computation nodes expose typed input and output ports.
- Edges carry immutable values produced within one logical reaction/epoch.
- State is represented canonically as a node, not as an edge annotation.
- Authoring tools may offer a feedback-edge shorthand, but it must desugar into a state node in the canonical IR.
- Every graph cycle must cross at least one state/delay node. A cycle composed only of combinational nodes is an invalid algebraic loop.

Conceptually:

```text
                  current / read
            +------------------------> decoder
            |                              |
       +----+-----+                        | update
       | KV state |<-----------------------+
       +----------+
            ^
         initial
```

State-node behavior is versioned:

```text
current[N] = committed_state[N]
pending[N+1] = update(current[N], delta[N])
```

A state input must never become visible through its output in the same reaction.

### State synchronization

Use synchronous reactions with snapshot isolation and two-phase state commit:

1. Snapshot all participating state at epoch `N`.
2. Execute nodes against that committed snapshot.
3. Stage writes into pending state for `N+1`.
4. Atomically commit the transaction group, or discard all pending writes on failure.

Required invariants:

- readers observe only committed state from their epoch;
- writers stage only the next version;
- one effective writer per state per reaction, unless an explicit merge/reduction defines the result;
- an untaken branch explicitly retains state rather than producing an implicit missing write;
- related KV, token-history, RNG, grammar, recurrent, and continuation state can commit atomically;
- synchronization is per sequence/transaction group, not a global barrier across unrelated requests;
- logical epochs must not require device-wide synchronization—GPU ordering should use streams and events.

The compiler should derive transaction groups from cyclic regions/SCCs and effect dependencies where possible. An explicit transaction-group declaration should be needed only when required atomicity cannot be inferred structurally.

## KV cache and paged attention

Portable metadata should describe logical KV semantics, not a physical storage policy:

- attention-state identity and scope;
- key/value roles and per-layer graph bindings;
- geometry and sequence axis;
- append/replace/indexed update semantics;
- aliasing requirements;
- snapshot, fork, commit, and token-granular rollback bounds.

The runtime owns:

- contiguous versus paged storage;
- page size and physical layout;
- block IDs, block tables, allocators, eviction, tiering, and offload;
- sequence slots and request/row identities.

The planner lowers the same logical state node to `ContiguousKvBuffer`, `PagedKvCache`, `SlidingWindowKvCache`, shared-prefix storage, or another compatible implementation. If a physical paged-attention kernel consumes `kv_blocks`, `block_table`, `context_lengths`, or `slot_mapping`, a runtime KV-state service generates those bindings during lowering; package authors do not serialize physical page tables.

Speculative execution should use the same transaction model: reserve and stage proposed positions, commit the accepted prefix, then roll back or release the remainder without copying the full cache.

## Compilation model

```text
authored node/edge graph
    -> parse and type-check
    -> SCC and algebraic-loop analysis
    -> state/effect/termination validation
    -> structured reactive regions
    -> resolved semantic state plan
    -> static execution plan
    -> runtime-specific physical lowering
```

The serialized graph remains the single source of truth. Structured regions, decoder ABI, state plans, capability requirements, cache dependencies, and physical execution plans are derived artifacts and must not be serialized as competing authorities.

Malformed or unsupported graphs should return diagnostics with node, port, edge, state identity, and dependency provenance rather than an ambiguous `None` or bare rejection.

## Performance contract

The abstraction must be compilable away before steady-state decode. Target invariants:

```text
graph interpretation in decode hot path = 0
extra tensor copies                   = 0
device-wide synchronization           = 0
per-token heap allocation              = 0
state commit                           = O(changed descriptors/pages)
```

The implementation must preserve continuous batching, paged attention, prefix reuse, fusion, static/direct dispatch, and CUDA graph capture. State versions should be handles/descriptors, not copied tensors. Paged-KV commit should normally update logical lengths and page references rather than materialize a new cache.

Performance cannot be inferred from the abstraction alone. Any prototype should be compared against the existing handwritten decode path using the same model, hardware, batch sizes, and context lengths, measuring TTFT, inter-token latency, throughput, host dispatch time, memory use, allocations, copies, and graph-capture/fallback counts while confirming identical generated token IDs.

## Questions to resolve

- What is the exact firing model: synchronous dataflow, token-driven execution, or another deterministic model?
- How are termination and inactive/gated values represented without reintroducing hidden control-flow semantics?
- Which state transaction groups can always be derived from SCCs and effect dependencies?
- Should general irreducible cyclic graphs be supported, or should the compiler accept only cycles that lower into recognized structured regions?
- What is the portable semantic type exposed by a runtime-managed state node or handle?
- How are component substitution and equivalence checked when a runtime replaces tensor KV ports with a paged-state kernel ABI?
- Should the initial prototype support only the single-decoder generation loop before generalizing to branches and nested/multiple feedback regions?

## Initial acceptance criteria

- Specify deterministic epoch, firing, state-delay, failure, and commit semantics.
- Define a canonical node/edge/state schema with no duplicate serialized graph ABI.
- Reject algebraic loops and ambiguous/multi-writer state with actionable diagnostics.
- Demonstrate lowering of a decoder + sampler + termination cycle.
- Demonstrate one logical KV graph lowering to both contiguous and paged implementations.
- Demonstrate atomic KV/token/RNG state rollback for a rejected speculative suffix.
- Show no graph interpretation, extra KV copies, device-wide synchronization, or per-token allocation in steady-state decode.
- Establish correctness and performance parity against the existing decode path before considering migration.

**Comment by justinchuby** [comment]
## Design checkpoint: synchronous reactive dataflow v8

This is a work-in-progress checkpoint from applying the proposed cyclic dataflow model to the published Qwen 2.5 package. It is not a final schema proposal yet.

### Core model

- The IR is **single-clock synchronous reactive dataflow**, not structured `sequence` / `branch` / `loop`.
- Components and generic scalar/bundled state are nodes.
- A component instance declares only inputs. Its outputs are the authoritative ONNX `ValueInfo` outputs, or the explicit signature of a runtime binding, and are referenced as `node_id.output_port`.
- ONNX dtype/rank/shape declarations are not repeated in YAML. Graph-visible ONNX I/O must have sufficient `ValueInfo` for static checking.
- Non-ONNX batching facts remain explicit: request-aligned axes, row selection/expansion, and other row-mapping behavior.
- A cycle is legal only when it crosses state. Removing state-delay edges must leave an instantaneous DAG.
- State carries no model-semantic `kind`. It declares only behavior that changes execution: transition, ownership, lifetime, aliasing, reuse, rollback, snapshot, and fork.
- State may be bundled. Qwen's 48 K/V tensors are one state node with 48 explicit members.
- An optional versioned interface may label canonical members to prove that a paged/native implementation substitution is legal. It cannot add or override graph edges.

### Execution shape

```mermaid
flowchart LR
    A["Admission baseline<br/>durable session state or initializer"] --> F{"reaction.first"}

    F -->|true| P["Prefill prompt<br/>against current KV"]
    F -->|false| C["Carried logits + KV"]

    P --> M["Presence merge<br/>exactly one source"]
    C --> M

    M --> S["Sample token"]
    S --> T["Termination + lengths"]
    S --> D["Decode sampled token"]
    D --> N["Stage next token, logits,<br/>KV, mask, RNG, control"]
    T --> N

    N --> W["Per-row reaction working commit"]
    W --> R{"repeat_while<br/>and i + 1 < max_reactions"}
    R -->|true| F
    R -->|false| X["state.final + clock.completed"]

    X --> O["Derived postlude<br/>for example VAE decode"]
    O --> Z["Invocation durable commit"]
```

The important distinction is:

1. A **reaction working commit** atomically advances one row's working state for the next reaction.
2. An **invocation durable commit** persists final session state, effects, and output heads only after the clock and postlude succeed.

Failure or cancellation can therefore restore the admission baseline even after several internal reactions. Early externally visible output requires the provisional-revision protocol and an `abort_to_baseline` outcome.

### Presence replaces authored branch structure

- A scalar `node.when` controls whether a node fires.
- A false gate produces an absent value, not zero, stale data, or implicit state retention.
- `bool[batch]` remains ordinary tensor data; it never implicitly changes graph firing or row compaction.
- Maybe-present values meet at an explicit typed `merge` requiring exactly one present input.
- Dynamic firing is valid. Every runtime must lower it to precompiled blocks/a finite state machine rather than interpret YAML in the hot path.

For Qwen, the compiler can produce one prefill block and one branch-free steady-state block:

```yaml
graph:
  clock:
    kind: synchronous
    max_reactions: request.max_iterations
    index: reaction.index
    first: reaction.first
    repeat_while: termination.continue
    sample_repeat: after_working_commit

  nodes:
    prefill:
      kind: component
      component: model
      when: {value: reaction.first, equals: true}
      inputs:
        input_ids: request.input_ids
        past_key_values.0.key: decoder_cache.current.key.0
        # ...

    carried_logits:
      kind: gate
      when: {value: reaction.first, equals: false}
      value: logits.current

    effective_logits:
      kind: merge
      inputs: [prefill_logits.last_logits, carried_logits.value]
      require: exactly_one_present

    sample:
      kind: component
      component: token_sampler
      inputs:
        logits: effective_logits.value
        counter: rng_counter.current
        active: active.current
        done: done.current

    decode:
      kind: component
      component: model
      inputs:
        input_ids: token_update.next
        past_key_values.0.key: effective_cache.value.key.0
        # ...

    logits:
      kind: state
      initial: absent
      next: decode_logits.last_logits
      transition: {kind: replace}
```

Initially absent state is allowed when presence analysis proves that reaction 0 cannot read it before the first successful write.

### State transitions

Candidate production and commit selection are separate:

```yaml
# Ordinary replacement
transition: {kind: replace}

# Linear speculative append
transition:
  kind: append
  axis: 2
  candidate_increment: verifier.evaluated
  commit: {kind: prefix, length: acceptance.length}
  bound: package.max_context

# Candidate tree / beam path selection
transition:
  kind: append
  axis: 2
  candidate_increment: tree.evaluated
  commit:
    kind: gather
    indices: acceptance.path_indices
    length: acceptance.length
  bound: package.max_context
```

Candidate trees remain ordinary tensor/component dataflow. State still has one candidate `next`; unselected candidates never become committed state, so v1 does not need first-class state-version fork nodes.

### Lifecycle without authored phases

Availability is inferred:

- `stable`: request/package inputs and pure closures that depend only on them;
- `reaction`: `reaction.index`, `state.current`, and dependent values;
- `final`: `state.final`, `clock.completed`, and dependent values.

This lets a diffusion VAE run once after the loop without an authored finalize region:

```yaml
vae:
  kind: component
  component: vae_decoder
  inputs:
    latent: latent.final
```

If `max_reactions` is zero, no reaction runs; initialized state becomes final immediately. For a Qwen package that must durably ingest a prompt, zero-output-budget behavior still needs an explicit package/admission rule.

### Output and effects

Each logical output stream has one binding to an ordered typed publication batch rather than multiple independent `emit` statements:

```yaml
outputs:
  tokens:
    publication:
      operations:
        - kind: append
          value: acceptance.tokens
          valid_length: acceptance.length
        - kind: append
          value: grammar.token
          valid_length: grammar.forced_length
```

Effects use typed linear tokens. Their commit behavior is independent from retry and speculation safety:

```yaml
effects:
  external_store:
    commit_behavior: transactional  # pure | transactional | after_invocation_commit
    retry: idempotent
    speculation_safety: {...}
```

`transactional` effects join working savepoints and final commit/abort. `after_invocation_commit` effects cannot influence state or the repeat predicate, and their failure cannot roll back durable state.

### Batching and performance

- Transactions are atomic per request-aligned row, not per physical batch.
- Mutable non-row-aligned state explicitly couples dependent rows.
- Stable nodes are hoisted and stable gates specialized at admission.
- Static reactions lower to a fixed prebound schedule and remain eligible for fusion/CUDA graph capture.
- Dynamic reactions lower to precompiled blocks/FSM transitions.
- The hot path has no YAML traversal, string/hash lookup, dependency discovery, or per-reaction heap allocation.
- State snapshots are handles/descriptors, not copied tensors.

Performance parity remains a measurement requirement: zero additional tensor copies, no device-wide synchronization across unrelated work, capture preserved where supported, and identical generated token IDs.

### Known gaps to resolve next

1. The concrete Qwen sketch still abbreviates several required policy nodes: termination configuration, generated/emitted length, cache length, and accepted length updates.
2. Reaction 0's cache candidate contains **prompt positions plus the sampled token**, so its `candidate_increment` and commit length cannot be the steady-state value `1`; the transition needs a first-reaction count derived from valid prompt length.
3. Session continuation must restore/advance cache logical lengths and mask state consistently with KV, not restore KV alone.
4. Zero-output-budget prompt persistence needs a deliberate package/admission rule.

The current direction is therefore coherent at the IR level, but the Qwen instance is not yet a complete replacement document.
